binding.cc

```diff
@@ -919,1155 +800,1353 @@ sn_crypto_stream_xor(
-
-  SN_ARGV_TYPEDARRAY(h, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-  SN_ARGV_TYPEDARRAY(k, 2)
-
-  SN_ASSERT_LENGTH(h_size, crypto_auth_BYTES, "h")
-  SN_ASSERT_LENGTH(k_size, crypto_auth_KEYBYTES, "k")
+static inline bool
+sn_crypto_auth_verify(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> h,
+  js_typedarray_span_t<> in,
+  js_typedarray_span_t<> k
+) {
+  assert(h.size_bytes() == crypto_auth_BYTES);
+  assert(k.size_bytes() == crypto_auth_KEYBYTES);
 
-  SN_RETURN_BOOLEAN(crypto_auth_verify(h_data, in_data, in_size, k_data))
+  return crypto_auth_verify(h.data(), in.data(), in.size_bytes(), k.data()) == 0;
 }
 
-js_value_t *
-sn_crypto_onetimeauth (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(3, crypto_onetimeauth)
-
-  SN_ARGV_TYPEDARRAY(out, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-  SN_ARGV_TYPEDARRAY(k, 2)
-
-  SN_ASSERT_LENGTH(out_size, crypto_onetimeauth_BYTES, "out")
-  SN_ASSERT_LENGTH(k_size, crypto_onetimeauth_KEYBYTES, "k")
+static inline int
+sn_crypto_onetimeauth(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> out,
+  js_typedarray_span_t<> in,
+  js_typedarray_span_t<> k
+) {
+  assert(out.size_bytes() == crypto_onetimeauth_BYTES);
+  assert(k.size_bytes() == crypto_onetimeauth_KEYBYTES);
 
-  SN_RETURN(crypto_onetimeauth(out_data, in_data, in_size, k_data), "failed to generate onetime authentication tag")
+  return crypto_onetimeauth(out.data(), in.data(), in.size_bytes(), k.data());
 }
 
-js_value_t *
-sn_crypto_onetimeauth_init (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_onetimeauth_init)
-
-  SN_ARGV_BUFFER_CAST(crypto_onetimeauth_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(k, 1)
+static inline int
+sn_crypto_onetimeauth_init(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> k
+) {
+  assert(state.size_bytes() == sizeof(crypto_onetimeauth_state));
+  auto state_data = reinterpret_cast<crypto_onetimeauth_state *>(state.data());
 
-  SN_THROWS(state_size != sizeof(crypto_onetimeauth_state), "state must be 'crypto_onetimeauth_STATEBYTES' bytes")
-  SN_ASSERT_LENGTH(k_size, crypto_onetimeauth_KEYBYTES, "k")
+  assert(k.size_bytes() == crypto_onetimeauth_KEYBYTES);
 
-  SN_RETURN(crypto_onetimeauth_init(state, k_data), "failed to initialise onetime authentication")
+  return crypto_onetimeauth_init(state_data, k.data());
 }
 
-js_value_t *
-sn_crypto_onetimeauth_update(js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_onetimeauth_update)
-
-  SN_ARGV_BUFFER_CAST(crypto_onetimeauth_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-
-  SN_THROWS(state_size != sizeof(crypto_onetimeauth_state), "state must be 'crypto_onetimeauth_STATEBYTES' bytes")
+static inline int
+sn_crypto_onetimeauth_update(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> in
+) {
+  assert(state.size_bytes() == sizeof(crypto_onetimeauth_state));
+  auto state_data = reinterpret_cast<crypto_onetimeauth_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.